### PR TITLE
Avoid meaningless expression evaluation in LayoutConstraintEvaluator

### PR DESCRIPTION
**Pull request by kokosing**
Avoid meaningless expression evaluation in LayoutConstraintEvaluator

When bindings do not provide any values used in expression there is no
point to evaluate such expression.

### CODE REVIEW

File: presto-tpch/src/main/java/com/facebook/presto/tpch/TpchConnectorFactory.java [hunk 1]
```diff
@@ -35,15 +35,22 @@
     public static final String TPCH_COLUMN_NAMING_PROPERTY = "tpch.column-naming";
 
     private final int defaultSplitsPerNode;
+    private final boolean predicatePushdownEnabled;
```
Comment (findepi) [inline presto-tpch/src/main/java/com/facebook/presto/tpch/TpchConnectorFactory.java]: Please rebase.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 2]
```diff
@@ -1331,10 +1333,16 @@ public LayoutConstraintEvaluator(Session session, Map<Symbol, Type> types, Map<S
...
 
             evaluator = ExpressionInterpreter.expressionOptimizer(expression, metadata, session, expressionTypes);
+            arguments = SymbolsExtractor.extractUnique(expression).stream()
+                    .map(assignments::get)
+                    .collect(toImmutableSet());
         }
 
         private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
         {
+            if (!arguments.stream().anyMatch(bindings.keySet()::contains)) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Isn't that `bindings.keySet()` will be the same in every call to `isCandidate()`?

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: for cases I am aware of it is true, but it is not enforced anywhere and connector may pass anything it want

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Shouldn't it be `allMatch` instead?

**Comment by kokosing** [comment]
Rebased

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 3]
```diff
@@ -1383,10 +1385,16 @@ public LayoutConstraintEvaluator(Session session, TypeProvider types, Map<Symbol
...
 
             evaluator = ExpressionInterpreter.expressionOptimizer(expression, metadata, session, expressionTypes);
+            arguments = SymbolsExtractor.extractUnique(expression).stream()
+                    .map(assignments::get)
+                    .collect(toImmutableSet());
         }
 
         private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
         {
+            if (!bindings.keySet().stream().anyMatch(arguments::contains)) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: use `intersection`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 4]
```diff
@@ -1397,5 +1406,10 @@ private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
 
             return true;
         }
+
+        private <T> boolean intersects(Set<T> one, Set<T> other)
+        {
+            return intersection(one, other).stream().findFirst().isPresent();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: `! intersection(one, other).isEmpty() `?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 5]
```diff
@@ -1397,5 +1406,10 @@ private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
 
             return true;
         }
+
+        private <T> boolean intersects(Set<T> one, Set<T> other)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: `one, other` -- this doesn't sound great.
`first, second` ? `a, b` ?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java (hunk 4 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: no,  `isEmpty()` is implemented with `size() == 0` I do not care the size

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: I think I used to use `first` and `second`... but I got comments like I should use `other` instead. I will use `first` and `second`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: Did you mean `com.google.common.collect.Sets.SetView#isEmpty` ?

**Pull request merged**